File: src/DirectInjection/Providers/ActionDependencyModelProvider.cs

using Microsoft.AspNetCore.Mvc.ApplicationModels;
using Microsoft.AspNetCore.Mvc.ModelBinding;

namespace DirectInjection.Providers {
    public class ActionDependencyModelProvider : IApplicationModelProvider {
        public int Order => -901;

        public void OnProvidersExecuted(ApplicationModelProviderContext context) {
        }

        public void OnProvidersExecuting(ApplicationModelProviderContext context) {
            foreach (var controller in context.Result.Controllers) {
                foreach (var action in controller.Actions) {
                    foreach (var parameter in action.Parameters) {
                        if (parameter.ParameterType.IsInterface) {
                            parameter.BindingInfo = new BindingInfo() {
                                BindingSource = BindingSource.Services
                            };
                        }
                    }
                }
            }
        }
    }
}

File: src/DirectInjection.FSharp/Providers/ActionDependencyModelProvider.fs
namespace Providers

open Microsoft.AspNetCore.Mvc.ApplicationModels
open Microsoft.AspNetCore.Mvc.ModelBinding

type ActionDependencyModelProvider() =
    interface IApplicationModelProvider with
        member __.Order = -901
        member __.OnProvidersExecuted _ = ()

        member __.OnProvidersExecuting (context: ApplicationModelProviderContext) =
            context.Result.Controllers
            |> Seq.collect (fun x -> x.Actions)
            |> Seq.collect (fun x -> x.Parameters)
            |> Seq.filter (fun x -> x.ParameterType.Name.EndsWith("Service"))
            |> Seq.iter (fun x -> x.BindingInfo <- BindingInfo(BindingSource = BindingSource.Services))